frappe/workflow/doctype/workflow_action/workflow_action.py

```diff
@@ -132,20 +138,111 @@ def return_link_expired_page(doc, doc_workflow_state):
 				frappe.bold(frappe.get_value('User', doc.get("modified_by"), 'full_name'))
 			), indicator_color='blue')
 
-def clear_old_workflow_actions(doc, user=None):
+
+def update_completed_workflow_actions(doc, user=None, workflow=None, workflow_state=None):
+	allowed_roles = get_allowed_roles(user, workflow, workflow_state)
+	# There is no transaction leading upto this state
+	# so no older actions to complete
+	if not allowed_roles:
+		return
+	if is_role_set(doc, allowed_roles):
+		update_completed_workflow_actions_using_role(doc, user, allowed_roles)
+		clear_old_workflow_actions_using_role(doc)
+	else:
+		# backwards compatibility
+		# for workflow actions saved using user
+		clear_old_workflow_actions_using_user(doc, user)
+		update_completed_workflow_actions_using_user(doc, user)
+
+def get_allowed_roles(user, workflow, workflow_state):
 	user = user if user else frappe.session.user
-	frappe.db.delete("Workflow Action", {
-		"reference_doctype": doc.get("doctype"),
-		"reference_name": doc.get("name"),
-		"user": ("!=", user),
-		"status": "Open"
-	})
 
-def update_completed_workflow_actions(doc, user=None):
+	allowed_roles = frappe.get_all('Workflow Transition',
+		fields='allowed',
+		filters=[['parent', '=', workflow],
+		['next_state', '=', workflow_state]],
+		pluck = 'allowed')
+
+	user_roles = set(frappe.get_roles(user))
+	return set(allowed_roles).intersection(user_roles)
+
+def is_role_set(doc, allowed_roles):
+	WorkflowAction = DocType("Workflow Action")
+	return frappe.qb.from_(WorkflowAction).select(WorkflowAction.status).where(
+			WorkflowAction.reference_doctype == doc.get('doctype')
+		).where(
+			WorkflowAction.reference_name == doc.get('name')
+		).where(
+			WorkflowAction.role.isin(list(allowed_roles))
+		).where(
+			WorkflowAction.status == 'Open',
+		).run()
+
+def update_completed_workflow_actions_using_role(doc, user=None, allowed_roles = set()):
 	user = user if user else frappe.session.user
-	frappe.db.sql("""UPDATE `tabWorkflow Action` SET `status`='Completed', `completed_by`=%s
-		WHERE `reference_doctype`=%s AND `reference_name`=%s AND `user`=%s AND `status`='Open'""",
-		(user, doc.get('doctype'), doc.get('name'), user))
+	WorkflowAction = DocType("Workflow Action")
+	# postgres requires order_by for limit to Work
+	frappe.qb.update(WorkflowAction).set(
+		WorkflowAction.status, 'Completed',
+	).set(
+		WorkflowAction.completed_by, user
+	).where(
+		WorkflowAction.reference_doctype == doc.get('doctype')
+	).where(
+		WorkflowAction.reference_name == doc.get('name')
+	).where(
+		WorkflowAction.role.isin(list(allowed_roles))
+	).where(
+		WorkflowAction.status == 'Open',
+	).where(
+		WorkflowAction.is_deleted == 0,
+	).orderby('role').limit(1).run()
```

Comment:
I don't think you can use ```orderyby``` in update, subquery or rowcount might be  better option.

Reply:
separated the select from update